### PR TITLE
Cherry-pick mismatched parameter optionality analyzer and clean up directories

### PR DESCRIPTION
**Pull request by captainsafia**
Cherry-picks https://github.com/dotnet/aspnetcore/pull/36154 to main and cleans up the folders in the Analyzers directory now that we have both components and minimal APIs analyzers in here.

### CODE REVIEW

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DetectMismatchedParameterOptionality.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.DelegateEndpoints;
+
+public partial class DelegateEndpointAnalyzer : DiagnosticAnalyzer
+{
+    internal const string DetectMismatchedParameterOptionalityRuleId = "ASP0006";
```
Comment (MackinnonBuck) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DetectMismatchedParameterOptionality.cs:16]: Should this be
```suggestion
    internal const string DetectMismatchedParameterOptionalityRuleId = "ASP0007";
```
?

File: src/Framework/Framework.slnf [hunk 2]
```diff
@@ -19,7 +19,8 @@
       "src\\Extensions\\Features\\src\\Microsoft.Extensions.Features.csproj",
       "src\\FileProviders\\Embedded\\src\\Microsoft.Extensions.FileProviders.Embedded.csproj",
       "src\\FileProviders\\Manifest.MSBuildTask\\src\\Microsoft.Extensions.FileProviders.Embedded.Manifest.Task.csproj",
-      "src\\Framework\\Analyzer\\src\\Microsoft.AspNetCore.App.Analyzers.csproj",
+      "src\\Framework\\AspNetCoreAnalyzers\\src\\Analyzers\\Microsoft.AspNetCore.App.Analyzers.csproj",
+      "src\\Framework\\AspNetCoreAnalyzers\\src\\CodeFixes\\Microsoft.AspNetCore.App.CodeFixes.csproj",
```
Comment (dougbu) [inline src/Framework/Framework.slnf:23]: Why aren't these in AspNetCore.sln❔ I don't think the solution filter will include these projects as-is.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DetectMismatchedParameterOptionality.cs (hunk 1 above)
Comment (MackinnonBuck) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DetectMismatchedParameterOptionality.cs:16]: Actually, is this used anywhere?

File: src/Framework/Framework.slnf (hunk 2 above)
Comment (dougbu) [reply src/Framework/Framework.slnf:23]: In addition, this file and AspNetCore.sln are missing the renamed src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj

These problems also exist in release/6.0

Comment (captainsafia) [reply src/Framework/Framework.slnf:23]: Ooof! I'll fix this up. I got a code-check warning about removing the old references but not one about adding the new ones.

Is there a chance we can add a code-check validation for that?

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DiagnosticDescriptors.cs [hunk 3]
```diff
@@ -43,4 +43,13 @@ internal static class DiagnosticDescriptors
...
         helpLinkUri: "https://aka.ms/aspnet/analyzers");
+
+    internal static readonly DiagnosticDescriptor DetectMismatchedParameterOptionality = new(
+        "ASP0007",
+        "Route parameter and argument optionality is mismatched",
+        "'{0}' argument should be annotated as optional or nullable to match route parameter",
+        "Usage",
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        helpLinkUri: "https://aka.ms/aspnet/analyzers");
```
Comment (MackinnonBuck) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DiagnosticDescriptors.cs:54]: Should this whole file be moved up one directory? It also contains the `DiagnosticDescriptor`s for `M.A.Components` analyzers.

File: src/Framework/Framework.slnf (hunk 2 above)
Comment (dougbu) [reply src/Framework/Framework.slnf:23]: > Is there a chance we can add a code-check validation for that?

Sure. I'll work on such a thing later today.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DetectMismatchedParameterOptionality.cs (hunk 1 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/DelegateEndpoints/DetectMismatchedParameterOptionality.cs]: Removed.

File: src/Framework/Framework.slnf (hunk 2 above)
Comment (captainsafia) [reply src/Framework/Framework.slnf:23]: Cool! Thanks. I'll submit a PR fixing this up in release/6.0

**Review by MackinnonBuck — APPROVED** [review]

File: src/Framework/Framework.slnf (hunk 2 above)
Comment (dougbu) [reply src/Framework/Framework.slnf:23]: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj is still not mentioned here nor in AspNetCore.sln

Comment (captainsafia) [reply src/Framework/Framework.slnf:23]: It's added in https://github.com/dotnet/aspnetcore/pull/36379/files#diff-d9f61b2475cc9938175a9c3e49e42bc4109ac983a29c73560932b2dc28fb4157R22 and https://github.com/dotnet/aspnetcore/pull/36379/files#diff-cd977e0a76b37d35c04d9d819ea66ef8a35d9ef7f86a9a7c774d751e8119db4fR1653.

Comment (dougbu) [reply src/Framework/Framework.slnf:23]: I keep copying the wrong project path ☹️ I meant src/Framework/AspNetCoreAnalyzers/**test**/Microsoft.AspNetCore.App.Analyzers.**Test**.csproj

File: src/Framework/Framework.slnf [hunk 4]
```diff
@@ -19,7 +19,8 @@
...
       "src\\Framework\\Analyzer\\test\\Microsoft.AspNetCore.App.Analyzers.Test.csproj",
```
Comment (dougbu) [inline src/Framework/Framework.slnf:24]: i.e.
```suggestion
      "src\\Framework\\AspNetCoreAnalyzers\\test\\Microsoft.AspNetCore.App.Analyzers.Test.csproj",
```

**Review by dougbu — COMMENTED** [review]
No further objections from me. Worst thing I could nitpick about (but won't 😀) is utility of trailing commas in array initializers to minimize perceived churn later. I'm not chiming in on approvals w/ @MackinnonBuck because I didn't really delve into the C#

**Comment by captainsafia** [comment]
@dougbu Sounds good. I think I'll need a approval from you since @MackinnonBuck has left the GitHub org.

**Review by dougbu — APPROVED** [review]

**Pull request merged**